packages/snfoundry/contracts/src/challenge0.cairo

```diff
@@ -35,28 +44,67 @@ mod Challenge0 {
         #[flat]
         ERC721Event: ERC721Component::Event,
         #[flat]
-        SRC5Event: SRC5Component::Event
+        SRC5Event: SRC5Component::Event,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event
     }
 
     #[constructor]
-    fn constructor(ref self: ContractState) {
+    fn constructor(ref self: ContractState, owner: ContractAddress) {
         let name: ByteArray = "YourCollectible";
         let symbol: ByteArray = "YCB";
-        let base_uri: ByteArray =
-            "https://ipfs.io/ipfs/QmfVMAmNM1kDEBYrC2TPzQDoCRFH6F5tE1e9Mr4FkkR5Xr"; // bison nft
+        let base_uri: ByteArray = "https://ipfs.io/";
 
         self.erc721.initializer(name, symbol, base_uri);
+        self.ownable.initializer(owner);
+    }
+
+    #[abi(embed_v0)]
+    impl WrappedIERC721MetadataImpl of IERC721Metadata<ContractState> {
+        fn token_uri(self: @ContractState, token_id: u256) -> ByteArray {
+            self._full_token_uri(token_id)
+        }
+        fn name(self: @ContractState) -> ByteArray {
+            self.erc721.name()
+        }
+        fn symbol(self: @ContractState) -> ByteArray {
+            self.erc721.symbol()
+        }
     }
 
     #[abi(embed_v0)]
-    pub impl Challenge0Impl of IChallenge0<ContractState> {
-        fn mint_item(ref self: ContractState, recipient: ContractAddress) -> u256 {
-            let id: u256 = 1;
-            self.erc721._mint(recipient, id); // _mint include _setTokenURI()
-            id
-        }
-        fn mint_id(ref self: ContractState, recipient: ContractAddress, id: u256) {
-            self.erc721._mint(recipient, id);
+    impl Challenge0Impl of IChallenge0<ContractState> {
+        fn mint_item(ref self: ContractState, recipient: ContractAddress, uri: ByteArray) -> u256 {
+            self._increment();
+            let token_id = self._current();
+            self.erc721._mint(recipient, token_id);
+            self._setTokenURI(token_id, uri);
+            token_id
+        }
+        fn tokenIdCounter(self: @ContractState) -> u256 {
+            self._current()
+        }
+    }
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn _increment(ref self: ContractState) {
+            self.counter.write(self.counter.read() + 1);
+        }
+
+        fn _current(self: @ContractState) -> u256 {
+            self.counter.read()
+        }
+
+        fn _full_token_uri(self: @ContractState, token_id: u256) -> ByteArray {
+            assert(self.erc721._exists(token_id), ERC721Component::Errors::INVALID_TOKEN_ID);
+            let base_uri = self.erc721._base_uri();
+            let token_uri = self.token_uris.read(token_id);
+            format!("{}{}", base_uri, token_uri)
+        }
+
+        fn _setTokenURI(ref self: ContractState, token_id: u256, uri: ByteArray) {
```

Comment:
keep all snake case, shouldnt be mixing unless explicitly using a `CamelOnlyImpl` implementation.